packages/dds/tree/src/test/simple-tree/schemaFactoryRecursive.spec.ts

```diff
@@ -0,0 +1,386 @@
+
+				type ExtractItemType2<Item extends () => unknown> = Item extends () => infer Result
+					? Result
+					: never;
+
+				type TFieldContentNode7 = ExtractItemType<TFieldContentNode3>;
+				type TFieldContentNode8 = ExtractItemType2<TFieldContentNode3>;
+				type TFieldContentNode9 = ReturnType<TFieldContentNode3>;
+				type X = typeof Box & TreeNodeSchema;
+				type XX = ReturnType<(() => 1) & (() => number)>;
+				type XY = ReturnType<(() => number) & (() => 1)>;
+
+				type Field = Box["child"];
+			}
+
+			const stuff: undefined | Box = view.root.child;
+
+			assert.equal(stuff, undefined);
+
+			view.root.child = new Box({
+				text: "hi2",
+				child: new Box({ text: "hi3", child: new Box({ text: "hi4", child: undefined }) }),
+			});
+
+			{
+				type _check1 = requireAssignableTo<undefined, typeof view.root.child.child>;
+				type _check2 = requireAssignableTo<Box, typeof view.root.child.child>;
+			}
+
+			const stuff2 = view.root.child?.child?.child;
+
+			assert.equal(stuff2?.text, "hi4");
+		});
+
+		it("objects", () => {
+			class ObjectRecursive extends sf.objectRecursive("Object", {
+				x: sf.optionalRecursive([() => ObjectRecursive]),
+			}) {}
+
+			type FieldsSchema = (typeof ObjectRecursive)["info"];
+			type XSchema = FieldsSchema["x"];
+			type Fields = ObjectFromSchemaRecordUnsafe<FieldsSchema>;
+			type Field = TreeFieldFromImplicitFieldUnsafe<XSchema>;
+			type Field2 = XSchema extends FieldSchema<infer Kind, infer Types>
+				? ApplyKind<TreeNodeFromImplicitAllowedTypes<Types>, Kind>
+				: "zzz";
+			type XTypes = XSchema extends FieldSchemaUnsafe<infer Kind, infer Types> ? Types : "Q";
+			type Field3 = TreeNodeFromImplicitAllowedTypes<XTypes>;
+			type Field4 = FlexListToUnion<XTypes>;
+
+			type Insertable = InsertableTreeNodeFromImplicitAllowedTypes<typeof ObjectRecursive>;
+			type Constructable = NodeFromSchema<typeof ObjectRecursive>;
+			type Child = ObjectRecursive["x"];
+			type _check = requireTrue<areSafelyAssignable<Child, ObjectRecursive | undefined>>;
+
+			const tree = hydrate(ObjectRecursive, new ObjectRecursive({ x: undefined }));
+
+			const data = Reflect.ownKeys(tree);
+			// TODO: are empty optional fields supposed to show up as keys in simple-tree? The currently are included, but maybe thats a bug?
+			assert.deepEqual(data, ["x"]);
+
+			tree.x = new ObjectRecursive({ x: undefined });
+
+			tree.x = tree.x?.x?.x?.x ?? new ObjectRecursive({ x: undefined });
+
+			const tree2 = hydrate(
+				ObjectRecursive,
+				new ObjectRecursive({ x: new ObjectRecursive({ x: undefined }) }),
+			);
+		});
+
+		it("other under recursive object", () => {
+			class Other extends sf.object("Other", {
+				y: sf.number,
+			}) {}
+			class ObjectRecursive extends sf.objectRecursive("Object", {
+				x: sf.optionalRecursive([() => ObjectRecursive]),
+				a: Other,
+				b: [Other],
+				c: [() => Other],
+				d: sf.optional(Other),
+				e: sf.optional([() => Other]),
+			}) {}
```

Comment:
Does recursion work without classes?

e.g., const myObj = _.objectRecursive(..., { a: [() => myObj] });

Reply:
seems to work as well as I'd expect: intellisense is rough, but it compiles. Any reason you ask?

Reply:
I believe that if you do that the generated d.ts files from the schema will end up with any in them making them screwed up if used across compilation boundaries. The initial reason we did class based schema was to fix that issue. This detail is documented on SchemaFactory (was added recently, see the table in its docs). I haven't revalidated this recently, but I do not expect it to be impacted by this PR.

Reply:
Ah, interesting. Probably doesn't apply here since the .d.ts file for this test is empty (no exports), but makes sense to not test a pattern which we don't support for customer use--people will generally want their schemas to produce well-formed types.

Reply:
I added an example showing this issue, see https://github.com/microsoft/FluidFramework/pull/19850/files#r1511847962